jax/interpreters/sharded_jit.py

```diff
@@ -204,6 +216,30 @@ def get_num_partitions(*partitions):
   return num_partitions_set.pop()
 
 
+def _inner_partitions(jaxpr, expected_num_parts: Optional[int]):
+  """Returns the total number of partitions from PartitonSpecs inside `jaxpr`.
+
+  Also validates that this number matches `expected_num_parts` if provided.
+  """
+  for eqn in jaxpr.eqns:
+    if eqn.primitive == set_sharding_p:
+      parts = eqn.params["partitions"]
+      nparts = get_num_partitions(parts)
+      if expected_num_parts is None:
+        expected_num_parts = nparts
+      elif nparts != expected_num_parts:
+        # TODO(skye): raise this error as we trace the jaxpr
```

Comment:
That definitely seems like better UX, but I'm also wary of bringing too much about pmap/sharding/etc into the JaxprTrace machinery. For maps, what I was thinking (very tentatively) was to create a final-style, dynamically-scoped "SPMD tracer" that maintains the axis context, propagates replication, handles axis_index, etc.; if you want the current pmap behavior, you can trace the output of that to an SPMD jaxpr (and translate that to SPMD XLA), but you can also wrap it in axis-consuming traces for other kinds of maps (vmap, soft_pmap, op-by-op pmap). Maybe something like that makes sense for partitions; maybe not. I guess all I'm saying is: I don't think we should necessarily worry about fixing this part now/soon 🙂 .

Reply:
Agreed we shouldn't bake any of this logic into the core tracing machinery. I like your tentative SPMD tracer idea. I was vaguely imagining using dynamic-scoped tracing and/or expanding pmap's DynamicAxisEnv concept to do final-style tracing for all parallel constructs, but it might be too complicated. I think we'll get more clarity as we expand the current initial-style transforms to compose with each other.